xml/encoder_test.go

```diff
@@ -0,0 +1,447 @@
+
+	if e, a := expect, encoder.String(); e != a {
+		t.Errorf("expect bodies to match, did not.\n,\tExpect:\n%s\n\tActual:\n%s\n", e, a)
+	}
+}
+
+func TestEncodeNamespace(t *testing.T) {
+	encoder := xml.NewEncoder()
+	obj := encoder.RootElement()
+
+	// nested `payload` shape
+	o := obj.Key("payload")
+	n, closeFn1 := o.NestedElement()
+
+	// nested `namespace` shape
+	n1 := n.Key("namespace", func(t *xml.TagMetadata) {
+		t.NamespacePrefix = "prefix"
+		t.NamespaceURI = "https://example.com"
+	})
+
+	// nested `prefixed` shape
+	n2, closeFn2 := n1.NestedElement()
+	n2.Key("prefixed").String("abc")
+
+	// call all close func's
+	closeFn2()
+	closeFn1()
+
+	e := []byte(`<payload><namespace xmlns:prefix="https://example.com"><prefixed>abc</prefixed></namespace></payload>`)
+	verify(t, encoder, e)
+}
+
+func verify(t *testing.T, encoder *xml.Encoder, e []byte) {
+	if a := encoder.Bytes(); bytes.Compare(e, a) != 0 {
+		t.Errorf("expected %+q, but got %+q", e, a)
+	}
+
+	if a := encoder.String(); string(e) != a {
+		t.Errorf("expected %s, but got %s", e, a)
+	}
+}
+
+func TestEncodeNestedShape(t *testing.T) {
+	encoder := xml.NewEncoder()
+	v := encoder.Value
+	root := v.RootElement()
+
+	// nested `payload` shape
+	o := root.Key("payload")
+
+	// nested `nested` shape
+	n1, closeFn1 := o.NestedElement()
+	o2 := n1.Key("nested")
+
+	// nested `value` shape
+	n2, closeFn2 := o2.NestedElement()
+	n2.Key("value").String("expected value")
+
+	// call all close fn's
+	closeFn2()
+	closeFn1()
+
+	e := []byte(`<payload><nested><value>expected value</value></nested></payload>`)
+	verify(t, encoder, e)
+}
+
+func TestEncodeMapString(t *testing.T) {
+	encoder := xml.NewEncoder()
+	v := encoder.Value
+	root := v.RootElement()
+
+	// nested `payload` shape
+	o := root.Key("payload")
+
+	// nested `mapStr` shape
+	ns, closeFn0 := o.NestedElement()
+	o1 := ns.Key("mapstr")
+
+	m, closeFn1 := o1.Map()
+
+	e, closeFn2 := m.Entry()
+	e.Key("key").String("abc")
+	e.Key("value").Integer(123)
+	closeFn2()
+
+	closeFn1()
+	closeFn0()
+
+	ex := []byte(`<payload><mapstr><entry><key>abc</key><value>123</value></entry></mapstr></payload>`)
+	verify(t, encoder, ex)
+}
+
+func TestEncodeMapFlatten(t *testing.T) {
+	encoder := xml.NewEncoder()
+	v := encoder.Value
+	root := v.RootElement()
+
+	// nested `payload` shape
+	o := root.Key("payload")
+
+	// nested `mapStr` shape
+	ns, closeFn0 := o.NestedElement()
+	o1 := ns.Key("mapstr")
+
+	m := o1.FlattenedMap()
+
+	e, closeFn2 := m.Entry()
+	e.Key("key").String("abc")
+	e.Key("value").Integer(123)
+	closeFn2()
+
+	closeFn0()
+
+	ex := []byte(`<payload><mapstr><key>abc</key><value>123</value></mapstr></payload>`)
+	verify(t, encoder, ex)
+}
+
+func TestEncodeMapNamed(t *testing.T) {
+	encoder := xml.NewEncoder()
+	v := encoder.Value
+	root := v.RootElement()
+
+	// nested `payload` shape
+	o := root.Key("payload")
+
+	// nested `mapStr` shape
+	ns, closeFn0 := o.NestedElement()
+	o1 := ns.Key("mapNamed")
```

Comment:
Kind of confusing for a RootElement to contain multiple keys, when I think the root Element it self should take the key name.  Might help usage and prevent bugs for RootElement to take the same parameters as Key to initialize the root element object and returns that. This would ensure that multiple XML documents are not rendered, and only one document is rendered per root element.